cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/worker/WorkerProgramRunner.java

```diff
@@ -105,7 +105,12 @@ public ProgramController run(Program program, ProgramOptions options) {
     int instanceId = Integer.parseInt(options.getArguments().getOption(ProgramOptionConstants.INSTANCE_ID, "-1"));
     Preconditions.checkArgument(instanceId >= 0, "Missing instance Id");
 
+    WorkerSpecification workerSpec = appSpec.getWorkers().get(program.getName());
+
     int instanceCount = Integer.parseInt(options.getArguments().getOption(ProgramOptionConstants.INSTANCES, "0"));
+    if (instanceCount == 0) {
+      instanceCount = Integer.parseInt(String.valueOf(workerSpec.getInstances()));
```

Comment:
does this mean the program options are not getting correctly set? Seems like it should be fixed by whatever it incorrectly setting it rather than getting it from the spec here?

Reply:
Hmm actually it should come from `KubeTwillContext`

Reply:
Had an offline discussion with @chtyim  and @albertshau 
instanceCount and instanceId should be coming from KubeContext. However, it looks like current KubeContext does not provide instanceId for non-statefulsets. 
We plan to implement compare and swap to get instanceId using kubernetes PUT api similar to implementing  leader election on k8s: 
https://kubernetes.io/blog/2016/01/simple-leader-election-with-kubernetes/#implementing-leader-election-in-kubernetes

Reply:
Consensus on instance ids based on above approach.

![image](https://user-images.githubusercontent.com/14131070/128098763-1192e30d-76c9-4976-9f5c-df44537e2c64.png)

![image](https://user-images.githubusercontent.com/14131070/128098866-d41bdf6c-ee60-425b-8dd7-56c53ed4196b.png)